core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java

```diff
@@ -230,14 +253,54 @@ protected void prepareOnAffectedNodes(TxInvocationContext ctx, PrepareCommand co
    }
 
    @Override
-   public Object visitRollbackCommand(TxInvocationContext ctx, RollbackCommand command) throws Throwable {
-      if (shouldInvokeRemoteTxCommand(ctx)) {
-         boolean syncRollback = cacheConfiguration.transaction().syncRollbackPhase();
-         ResponseMode responseMode = syncRollback ? ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS : ResponseMode.ASYNCHRONOUS;
-         rpcManager.invokeRemotely(getCommitNodes(ctx), command, rpcManager.getRpcOptionsBuilder(responseMode, DeliverOrder.NONE).build());
+   protected boolean needValuesFromPreviousOwners(InvocationContext ctx, WriteCommand command) {
+      if (ctx.isOriginLocal()) {
+         // The return value only matters on the originator.
+         // Conditional commands also check the previous value only on the originator.
+         if (isNeedReliableReturnValues(command) || command.isConditional())
+            return true;
       }
+      return !command.hasFlag(Flag.CACHE_MODE_LOCAL) && (shouldFetchRemoteValuesForWriteSkewCheck(ctx, command) || command.hasFlag(Flag.DELTA_WRITE));
+   }
 
-      return invokeNextInterceptor(ctx, command);
+   protected void remoteGetBeforeWrite(InvocationContext ctx, WriteCommand command, RecipientGenerator keygen) throws Throwable {
+      for (Object k : keygen.getKeys()) {
+         CacheEntry entry = ctx.lookupEntry(k);
+         boolean skipRemoteGet = entry != null && entry.skipLookup();
+         if (skipRemoteGet) {
+            continue;
+         }
+         InternalCacheEntry ice = remoteGet(ctx, k, true, command);
+         if (ice == null) {
+            localGet(ctx, k, true, command, false);
+         }
+      }
+   }
+
+   private Object visitGetCommand(InvocationContext ctx, AbstractDataCommand command,
+                                  boolean isGetCacheEntry) throws Throwable {
+      Object returnValue = invokeNextInterceptor(ctx, command);
+
+      //if the cache entry has the value lock flag set, skip the remote get.
+      CacheEntry entry = ctx.lookupEntry(command.getKey());
+      boolean skipRemoteGet = entry != null && entry.skipLookup();
+
+      // need to check in the context as well since a null retval is not necessarily an indication of the entry not being
+      // available.  It could just have been removed in the same tx beforehand.  Also don't bother with a remote get if
+      // the entry is mapped to the local node.
```

Comment:
Reading this now, it seems really wrong, i.e. `GetKeyValueCommand.perform()` should always return the right value, and it should be the distribution interceptor's job to put the value in the context so that the command can do its job. But I guess it's not related to this PR :)